### PR TITLE
[FIX] account: reconciliation on reversed move

### PR DESCRIPTION
**Pull request by guva-odoo**
With this commit, we make sure that all lines are fully reconciled when
reversing a move.

Steps to reproduce:

- Create a Journal Entry, with eg a line with 300$ debit/credit and two others
lines to make the journal entry fully balanced

|   Account   |   Partner   |   Debit   |   Credit   |
|-------------|-------------|-----------|------------|
|  Account X  |             |   300.0   |     0.0    |
|  Account Y  |  partner A  |     0.0   |   100.0    |
|  Account Y  |  partner B  |     0.0   |   200.0    |

- Then reverse the move
-> Only one line is marked as fully reconciled, the other one is marked
as partially reconciled.

This is because each line was passed in the reconcile method with
all the counterpart lines, even those which didn't belong to it.
Therefore, all the firsts lines was marked as partially reconciled
until the last one, which passed with the only counterpart line
left.

With this commit, we group the lines and counterpart lines by account,
and reconcile them.

This is a correction of this commit 6481eb720f36f3e81c2ec8371cccdd893a5e8ddf
which introduced a bug in l10n_mx, when unreconciling moves with tax 0%,
here are the steps to reproduce this bug:

- With l10n_mx installed
- Create an invoice with tax 0%
- Confirm and register payment
- Try to unreconcile the payment
-> We get an UserError : "You are trying to reconcile some entries that
  are already reconciled"

This happened because in the first fix, we didn't filtered the already
reconciled moves. Also, it's reproductible in 15 but change is made here
in order to keep consistency between versions.

opw-2810392
opw-2851999

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by Levizar — COMMENTED** [review]
@william-andre 
It's solving the mexican issue (when you unreconcile a payment from an invoice, it reverses it and try to reconcile but get a traceback because it is already reconciled) by just achieving a partial instead of a full_reconcile.
Would it be ok to add `abs(balance)` in the key so we would have the 1-1 matching most of the time?

**Review by william-andre — CHANGES_REQUESTED** [review]
Wouldn't the right fix be to unreconcile the payment before cancelling it?

> Create a Journal Entry, with eg a line with 300$ debit/credit and two others lines to make the journal entry fully balanced

This is very, very vague. What would the move look like at the end?

> which introduced a bug in l10n_mx.

I'm not sure how anything in this commit is related to `l10n_mx`. You don't explain what the bug is, but you give reproduction steps without localization specific steps?

> opw-2810392

This seems to have been fixed, without any followup on the chatter about something broken

**Comment by guva-odoo** [comment]
@william-andre @Levizar 
The bug is happening only since v15, and it’s related to this commit 65b5ec4ad65f798b1d647978b5b42a49e760139a.
In l10n_mx IVA 0% is linked to an account which allows reconciliation, 
since the commit mentioned we generate a line for tax with 0 amount which, for some reason I didn’t manage to find, is marked as reconciled after the call of move.write() https://github.com/odoo/odoo/blob/a5ee03cabc0fbd70d78e3e93b4ff001bb2861379/addons/account/models/account_move.py#L2292
However, I tried to do `(move.line_ids + reverse_move.line_ids)filtered(‘reconciled’).remove_move_reconcile()` here https://github.com/odoo/odoo/blob/a5ee03cabc0fbd70d78e3e93b4ff001bb2861379/addons/account/models/account_move.py#L2795 but as the line has no matched_debit_ids/matched_credit_ids, it stays reconciled.
So I guess the best solution is to filtered the reconciled lines as we already did before the fix.

**Comment by guva-odoo** [comment]
Hello @william-andre do you think we can merge it before Monday ? There is a batch of mx clients that are waiting for a solution. If we don't merge this now, we should at least revert the guilty fix, there is a pr here https://github.com/odoo/odoo/pull/91792 but it has been closed and I don't have rights to reopen it, maybe you have ?

**Review by william-andre — COMMENTED** [review]
It is not trivial to determine when we should unreconcile before (fully) reverting the move, so lets keep this fix for now...

@robodoo r+

**Comment by yelizariev** [comment]
Also, opw-2854731

**Comment by tsb-odoo** [comment]
@william-andre This fix had side-effects. OPW-2860581

**Comment by guva-odoo** [comment]
@tsb-odoo this is the side effect this commit is supposed to fix

**Comment by william-andre** [comment]
@tsb-odoo Most of the comments on the OPW are from before this got merged.
It only got merged on Monday at noon this week, so I would not expect it to be on production yet.

**Pull request closed without merging**
